api/metric/api.go

```diff
@@ -21,62 +23,231 @@ import (
 	"go.opentelemetry.io/api/unit"
 )
 
-type MetricType int
+// Kind categorizes different kinds of metric.
+type Kind int
 
 const (
-	Invalid    MetricType = iota
-	Gauge                 // Supports Set()
-	Cumulative            // Supports Inc()
+	Invalid     Kind = iota
+	CounterKind      // Supports Add()
+	GaugeKind        // Supports Set()
+	MeasureKind      // Supports Record()
+	ObserverKind
 )
 
-type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Handle interface {
+	// Record allows the SDK to observe a single metric event
+	RecordOne(ctx context.Context, value MeasurementValue)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+// LabelSet represents a []core.KeyValue for use as pre-defined labels
+// in the metrics API.
+//
+// TODO this belongs outside the metrics API, in some sense, but that
+// might create a dependency.  Putting this here means we can't re-use
+// a LabelSet between metrics and tracing, even when they are the same
+// SDK.
+type LabelSet interface {
+	Meter() Meter
 }
 
-type Handle struct {
-	Name        string
-	Description string
-	Unit        unit.Unit
+// ObserverCallback defines a type of the callback SDK will call for
+// the registered observers.
+type ObserverCallback func(Meter, Observer) (LabelSet, MeasurementValue)
+
+// Meter is an interface to the metrics portion of the OpenTelemetry SDK.
+type Meter interface {
+	// DefineLabels returns a reference to a set of labels that
+	// cannot be read by the application.
+	DefineLabels(context.Context, ...core.KeyValue) LabelSet
+
+	NewHandle(context.Context, Descriptor, LabelSet) Handle
+	DeleteHandle(Handle)
+
+	// RecordBatch atomically records a batch of measurements..
+	RecordBatch(context.Context, LabelSet, ...Measurement)
+
+	RegisterObserver(Observer, ObserverCallback)
+	UnregisterObserver(Observer)
+}
+
+type DescriptorID uint64
+type MetricValueKind int8
+
+const (
+	AnyValueKind MetricValueKind = iota
+	Int64ValueKind
+	Float64ValueKind
+)
+
+// Descriptor represents a named metric with recommended local-aggregation keys.
+type Descriptor struct {
```

Comment:
As I've worked on a prototype SDK now, I've come to think that Descriptors should not export any fields, and that the `New` operations should use `*Descriptor` pointers instead of structs. Inside the SDK, we end up needing to refer to a bunch of these fields, and presently this forces me to copy the Descriptor potentially many times in live memory. We could replace the exported fields with getter methods.

Reply:
The `New` methods take some option items (`WithDescription` and company) that set fields in the descriptor. Do we add setters for those, or keep the fields exported, or make the field unexported (thus ban options from other packages as they wouldn't be able to set anything in the descriptor).

Reply:
Maybe the setters could be `internal`? That would work.

Reply:
You'll see how #172 will benefit from this, currently it's copying the Descriptor into every record.

Reply:
For now I have unexported the fields and added the getters. This restricts options to the package, but I think I'm fine with it anyway. It's in [my branch](https://github.com/kinvolk/opentelemetry-go/tree/krnowak/metrics) for now.